### PR TITLE
bump bottle to version 0.14-dev

### LINKED ISSUE #604: Please update vendored `bottle`
[bottle](https://pypi.org/project/bottle/) 0.13.1 has been released recently, with support for Python 3.13. Unfortunately, from a quick look it introduced also some API changes.

### PR DESCRIPTION
**Pull request by oz123**
Support modern python versions.

This is commit 7af7135965e6e02993fea89d8932a4bc713b6692 from bottle's master branch.

Replace #546 , #613 #602 .
Fix #604.

### CODE REVIEW

**Comment by dee-me-tree-or-love** [comment]
Great! Thanks a lot @oz123 for fixing this one. 💟 

I'll approve and merge and will try to make sure `bottle` is treated like a regular dependency in the near future.

**Review by dee-me-tree-or-love — DISMISSED** [review]

**Comment by oz123** [comment]
I wonder why the test suite fails in a collosal fashion. I ran the tests locally with pytest and python 3.12.

**Comment by dee-me-tree-or-love** [comment]
> I wonder why the test suite fails in a collosal fashion. I ran the tests locally with pytest and python 3.12.

No worries! I am also looking into it right now, that's very interesting, I was also sure it will go pretty smooth 😄 
I guess good to not rush it now, but it's not a big problem. 

I'll keep you posted if I find anything in the coming hour, and looking forward to any updates if you're looking into it too!

**Comment by dee-me-tree-or-love** [comment]
@oz123 just a quick thought in the meantime, what do you think of upgrading `bottle` to the latest stable version? 
- https://github.com/bottlepy/bottle/releases/tag/0.13.2 

Then we're for sure relying on a production-ready version. WDYT?

**Comment by dee-me-tree-or-love** [comment]
> I managed to find the solution to one of the failing tests, `test_auto_servers`:

The other 4 failing tests: 

```
FAILED tests/test_app.py::test_nonroot_root - webtest.app.AppError: Bad response: 500 Internal Server Error (not 200 OK or 3xx redirect for http://nonr...
FAILED tests/test_app.py::test_nonroot_simple_index - webtest.app.AppError: Bad response: 500 Internal Server Error (not 200 OK or 3xx redirect for http://loca...
FAILED tests/test_app.py::test_nonroot_simple_packages - webtest.app.AppError: Bad response: 500 Internal Server Error (not 200 OK or 3xx redirect for http://loca...
FAILED tests/test_app.py::test_root_no_relative_paths - webtest.app.AppError: Bad response: 500 Internal Server Error (not 200 OK or 3xx redirect for http://loca...
```

All seem to have an issue with `Bottle.mount` that seems to have changed significantly in `0.13.+` 🤔  

[The documentation of `0.13.x`](https://bottlepy.org/docs/0.13/api.html#bottle.Bottle.mount) says: 

> Plugins from the parent application are not applied to the routes of the mounted child application. If you need plugins in the child application, install them separately.

While [older `0.12.x` docs](https://bottlepy.org/docs/0.12/api.html#bottle.Bottle.mount) say the following: 

> All other parameters are passed to the underlying [route()](https://bottlepy.org/docs/0.12/api.html#bottle.route) call.

In `test_app` we [define a `testpriv` fixture](https://github.com/pypiserver/pypiserver/blob/4ff5c7f1ff8240da5b20208ac19f589e53453c4a/tests/test_app.py#L50-L59) using the `Bottle.mount`. 

But in `app` we define a [couple of hooks](https://github.com/pypiserver/pypiserver/blob/4ff5c7f1ff8240da5b20208ac19f589e53453c4a/tests/test_app.py#L50-L59) that play a critical role in how the requests are handled.  

I don't have that confirmed, but it seems that the hooks are no longer working properly on a mounted app. 🤔

**Comment by oz123** [comment]
While you wrote your message, I also found the same solution.
The version included here is a bit newer than 0.13.2 and already removes alot of hacks for Python2.
Generally speaking, bottle is really rock solid, so I don't think it should be a problem to include something from the master branch. 
Marcel Hellkamp is quite conservative, which is also a good think in software. It is going to take a while until we see a 0.14 version, but I don't see a good reason why pypiserver should ship code that supports python2 if it has no intent to support Python2.

**Comment by oz123** [comment]
Regarding the hooks. I decided to remove the 'before-request' hook, because the actual access to the property it produced was only used sphoradically. The function call isn't more expensive, and isn't more verbose, so it seemed like a resonable replacement.

**Comment by dee-me-tree-or-love** [comment]
> While you wrote your message, I also found the same solution. 

Awesome! That's really great, thanks for working on this still 👍 
Finding the same solution is a good sign 😁  

> The version included here is a bit newer than 0.13.2 and already removes alot of hacks for Python2. Generally speaking, bottle is really rock solid, so I don't think it should be a problem to include something from the master branch. Marcel Hellkamp is quite conservative, which is also a good think in software. It is going to take a while until we see a 0.14 version, but I don't see a good reason why pypiserver should ship code that supports python2 if it has no intent to support Python2.

Fully convinced, good points!

File: pypiserver/_app.py [hunk 1]
```diff
@@ -90,7 +88,8 @@ def favicon():
 
 @app.route("/")
 def root():
-    fp = request.custom_fullpath
+    parsed = urlparse(request.urlparts.scheme + "://" + request.urlparts.netloc)
+    fp = parsed.path.rstrip("/") + "/" + request.fullpath.lstrip("/")
```
Comment (dee-me-tree-or-love) [inline pypiserver/_app.py:92]: Nice! Totally agree with replacing the hook with a direct function call. Just a reminder, I think you intended to replace this with `fp = request_fullpath(request)` too? 

```suggestion
    fp = request_fullpath(request)
```

**Comment by oz123** [comment]
One (hopefully) final note, I wanted to test locally with tox as the CI does. To do it I needed to install a few outdated Python versions. It not a big deal with Pyenv. But, instead I decided to drop old versions of Python.
Quite a scope creep for the PR's title. Sorry about that.

File: tox.ini [hunk 2]
```diff
@@ -1,5 +1,5 @@
 [tox]
-envlist = py36, py37, py38, py39, py310, py311, pypy3
+envlist = py310, py311, py312, pypy3
```
Comment (dee-me-tree-or-love) [inline tox.ini:2]: Oh good point! Thanks for updating this 👍 
I indeed see `py36` could have been removed already a while back 😅 

Maybe that's too picky from my side, but I feel a bit more comfortable keeping `py37, ... py39` for now, do you mind still including these? 

```suggestion
envlist = py37, py38, py39, py310, py311, py312, pypy3
```

According to [the `pypistats`](https://pypistats.org/packages/pypiserver) we are still getting some (albeit not much) downloads from these Python versions. I'd like to remove the `tox` configuration for them separately.

**Comment by dee-me-tree-or-love** [comment]
> One (hopefully) final note, I wanted to test locally with tox as the CI does. To do it I needed to install a few outdated Python versions. It not a big deal with Pyenv. But, instead I decided to drop old versions of Python. Quite a scope creep for the PR's title. Sorry about that.

That's totally good! Thank you for taking the time and putting effort into this. I feel our test setup and all the utils would benefit from some brushing up indeed 🫣 I'd like to address these separately at some point and it's a great insight in what's causing challenges. Thanks for thoroughly going through all of this

File: pypiserver/_app.py [hunk 3]
```diff
@@ -53,15 +58,8 @@ def protector(*args, **kwargs):
...
-
-@app.hook("before_request")
-def print_request():
-    parsed = urlparse(request.urlparts.scheme + "://" + request.urlparts.netloc)
-    request.custom_host = parsed.netloc
-    request.custom_fullpath = (
-        parsed.path.rstrip("/") + "/" + request.fullpath.lstrip("/")
-    )
+    request.custom_host = urlparse(
+            f"{request.urlparts.scheme}://{request.urlparts.netloc}")
```
Comment (dee-me-tree-or-love) [inline pypiserver/_app.py:62]: I see here we used to have `request.custom_host = parsed.netlock` but I actually don't see any references to `.custom_host` anywhere in the code. Maybe this could also go 🤔 .

Comment (dee-me-tree-or-love) [reply pypiserver/_app.py:62]: image

File: pypiserver/_app.py [hunk 4]
```diff
@@ -53,15 +58,8 @@ def protector(*args, **kwargs):
 @app.hook("before_request")
 def log_request():
     log.info(config.log_req_frmt, request.environ)
-
-
-@app.hook("before_request")
-def print_request():
-    parsed = urlparse(request.urlparts.scheme + "://" + request.urlparts.netloc)
-    request.custom_host = parsed.netloc
-    request.custom_fullpath = (
```
Comment (dee-me-tree-or-love) [inline pypiserver/_app.py:62]: I see the new tests might be failing on [`custom_fullpath` access in `core.py:L17`](https://github.com/pypiserver/pypiserver/blob/cf424c982d8c66b8ee38984b4c61b8340ddece28/pypiserver/core.py#L15-L17)

Comment (dee-me-tree-or-love) [reply pypiserver/_app.py:62]: @oz123 by moving `` to `_app.py` (and adjusting tests accordingly) like this: 

```
# Already added to `_app.py`
def request_fullpath(request):
    parsed = urlparse(request.urlparts.scheme + "://" + request.urlparts.netloc)
    return parsed.path.rstrip("/") + "/" + request.fullpath.lstrip("/")

# Moved from `core.py`
def get_bad_url_redirect_path(request, project):
    """Get the path for a bad root url."""
    uri = request_fullpath(request)
    if uri.endswith("/"):
        uri = uri[:-1]
    uri = uri.rsplit("/", 1)[0]
    project = quote(project)
    uri += f"/simple/{project}/"
    return uri
```

I think that should fix it 🤔

**Comment by dee-me-tree-or-love** [comment]
hey @oz123 I think I found the final workarounds: 

```bash
$ pytest ./tests
...
==================================== 492 passed, 1239 warnings in 20.61s =====================================
```


<details><summary> here's the <code>git diff</code> off your branch:</summary>

```
diff --git a/pypiserver/_app.py b/pypiserver/_app.py
index c334e30..e775dc7 100644
--- a/pypiserver/_app.py
+++ b/pypiserver/_app.py
@@ -9,11 +9,10 @@ from collections import defaultdict
 from collections import namedtuple
 from io import BytesIO
 from json import dumps
-from urllib.parse import urljoin, urlparse
+from urllib.parse import urljoin, urlparse, quote
 
 from pypiserver.config import RunConfig
 from . import __version__
-from . import core
 from .bottle import (
     static_file,
     redirect,
@@ -35,6 +34,16 @@ def request_fullpath(request):
     return parsed.path.rstrip("/") + "/" + request.fullpath.lstrip("/")
 
 
+def get_bad_url_redirect_path(uri, project):
+    """Get the path for a bad root url."""
+    if uri.endswith("/"):
+        uri = uri[:-1]
+    uri = uri.rsplit("/", 1)[0]
+    project = quote(project)
+    uri += f"/simple/{project}/"
+    return uri
+
+
 class auth:
     """decorator to apply authentication if specified for the decorated method & action"""
 
@@ -59,7 +68,8 @@ class auth:
 def log_request():
     log.info(config.log_req_frmt, request.environ)
     request.custom_host = urlparse(
-            f"{request.urlparts.scheme}://{request.urlparts.netloc}")
+        f"{request.urlparts.scheme}://{request.urlparts.netloc}"
+    ).netloc
 
 
 @app.hook("after_request")
@@ -146,10 +156,7 @@ def file_upload():
     )
     if not ufiles.pkg:
         raise HTTPError(400, "Missing 'content' file-field!")
-    if (
-        ufiles.sig
-        and f"{ufiles.pkg.raw_filename}.asc" != ufiles.sig.raw_filename
-    ):
+    if ufiles.sig and f"{ufiles.pkg.raw_filename}.asc" != ufiles.sig.raw_filename:
         raise HTTPError(
             400,
             f"Unrelated signature {ufiles.sig!r} for package {ufiles.pkg!r}!",
@@ -219,17 +226,11 @@ def handle_rpc():
     """Handle pip-style RPC2 search requests"""
     parser = xml.dom.minidom.parse(request.body)
     methodname = (
-        parser.getElementsByTagName("methodName")[0]
-        .childNodes[0]
-        .wholeText.strip()
+        parser.getElementsByTagName("methodName")[0].childNodes[0].wholeText.strip()
     )
     log.debug(f"Processing RPC2 request for '{methodname}'")
     if methodname == "search":
-        value = (
-            parser.getElementsByTagName("string")[0]
-            .childNodes[0]
-            .wholeText.strip()
-        )
+        value = parser.getElementsByTagName("string")[0].childNodes[0].wholeText.strip()
         response = []
         ordering = 0
         for p in config.backend.get_all_packages():
@@ -244,9 +245,7 @@ def handle_rpc():
                 }
                 response.append(d)
             ordering += 1
-        call_string = xmlrpclib.dumps(
-            (response,), "search", methodresponse=True
-        )
+        call_string = xmlrpclib.dumps((response,), "search", methodresponse=True)
         return call_string
 
 
@@ -326,9 +325,7 @@ def list_packages():
         key=lambda x: (os.path.dirname(x.relfn), x.pkgname, x.parsed_version),
     )
 
-    links = (
-        (pkg.relfn_unix, urljoin(fp, pkg.fname_and_hash)) for pkg in packages
-    )
+    links = ((pkg.relfn_unix, urljoin(fp, pkg.fname_and_hash)) for pkg in packages)
 
     tmpl = """<!DOCTYPE html>
 <html lang="en">
@@ -403,4 +400,5 @@ def json_info(project):
 @app.route("/:project/")
 def bad_url(project):
     """Redirect unknown root URLs to /simple/."""
-    return redirect(core.get_bad_url_redirect_path(request, project))
+    uri = request_fullpath(request)
+    return redirect(get_bad_url_redirect_path(uri, project))
diff --git a/pypiserver/core.py b/pypiserver/core.py
index df548ab..651de53 100644
--- a/pypiserver/core.py
+++ b/pypiserver/core.py
@@ -3,7 +3,6 @@
 
 import mimetypes
 import typing as t
-from urllib.parse import quote
 
 from pypiserver.pkg_helpers import normalize_pkgname, parse_version
 
@@ -12,17 +11,6 @@ mimetypes.add_type("application/octet-stream", ".whl")
 mimetypes.add_type("text/plain", ".asc")
 
 
-def get_bad_url_redirect_path(request, project):
-    """Get the path for a bad root url."""
-    uri = request.custom_fullpath
-    if uri.endswith("/"):
-        uri = uri[:-1]
-    uri = uri.rsplit("/", 1)[0]
-    project = quote(project)
-    uri += f"/simple/{project}/"
-    return uri
-
-
 class PkgFile:
     __slots__ = [
         "pkgname",  # The projects/package name with possible capitalization
diff --git a/tests/test_app.py b/tests/test_app.py
index 6471f6b..6e1fc69 100644
--- a/tests/test_app.py
+++ b/tests/test_app.py
@@ -13,7 +13,7 @@ import webtest
 
 # Local Imports
 from tests.test_pkg_helpers import files, invalid_files
-from pypiserver import __main__, bottle, core, Bottle
+from pypiserver import __main__, bottle, core, Bottle, _app
 from pypiserver.backend import CachingFileBackend, SimpleFileBackend
 
 # Enable logging to detect any problems with it
@@ -116,6 +116,14 @@ def add_file_to_root(app):
     return file_adder
 
 
+def test_redirect_project_encodes_newlines():
+    """Ensure raw newlines are url encoded in the generated redirect."""
+    uri = "/\nSet-Cookie:malicious=1;"
+    project = "\nSet-Cookie:malicious=1;"
+    newpath = _app.get_bad_url_redirect_path(uri, project)
+    assert "\n" not in newpath
+
+
 def test_root_count(root, testapp, add_file_to_root):
     """Test that the welcome page count updates with added packages
 
diff --git a/tests/test_core.py b/tests/test_core.py
index d74710d..38586f2 100644
--- a/tests/test_core.py
+++ b/tests/test_core.py
@@ -59,14 +59,6 @@ def test_fname_and_hash(tmp_path, hash_algo):
     assert pkgfile.fname_and_hash == f"{f.name}#{digester(pkgfile)}"
 
 
-def test_redirect_project_encodes_newlines():
-    """Ensure raw newlines are url encoded in the generated redirect."""
-    request = Namespace(custom_fullpath="/\nSet-Cookie:malicious=1;")
-    project = "\nSet-Cookie:malicious=1;"
-    newpath = core.get_bad_url_redirect_path(request, project)
-    assert "\n" not in newpath
-
-
 def test_normalize_pkgname_for_url_encodes_newlines():
     """Ensure newlines are url encoded in package names for urls."""
     assert "\n" not in normalize_pkgname_for_url("/\nSet-Cookie:malicious=1;")
```

</details>

**Comment by oz123** [comment]
Thanks for your engagement and quick responses! I pushed a few more commits adding all of your changes and some more fixes.
The tests should pass now.

**Comment by dee-me-tree-or-love** [comment]
> Thanks for your engagement and quick responses! I pushed a few more commits adding all of your changes and some more fixes. The tests should pass now.

Awesome! Thanks a lot @oz123, it was my pleasure, thanks a lot for putting your effort and time into this! 👍 

I see the tests are running well now ✔️ 
There's a small code-formatting issue spotted by `black`, if you could fix this up, I'll be super happy to merge.

**Comment by oz123** [comment]
Fixed that issue with formatting.

**Review by dee-me-tree-or-love — APPROVED** [review]
This is superb, thanks a lot for the effort @oz123, great job! 
Happy to merge this now 👍

**Comment by harryzcy** [comment]
@dee-me-tree-or-love Can you cut a new release with this change?

**Comment by dee-me-tree-or-love** [comment]
> @dee-me-tree-or-love Can you cut a new release with this change?

Hello @harryzcy, yes, thank you for the bump, I will be releasing it this weekend 👍

**Comment by dee-me-tree-or-love** [comment]
@harryzcy, a friendly bump, it should now be included in the [latest `v2.3.1` release](https://github.com/pypiserver/pypiserver/releases/tag/v2.3.1). 
Hopefully everything works out well for you. Be welcome to report any issues or leave comments :)

File: setup.py [hunk 5]
```diff
@@ -22,7 +22,7 @@
 install_requires = [
     "pip>=7",
     "packaging>=23.2",
-    "importlib_resources;python_version>'3.8' and python_version<'3.12'",
```
Comment (mgorny) [inline setup.py:25]: Why did you remove the Python version specifier? FWICS the code is still unchanged, and built-in `importlib.resources.files()` are used on Python 3.12.

Comment (oz123) [reply setup.py:25]: I removed the constraint smaller than Python 3.12 because the code is working for this Python version. 
Did you find and issue with this version?

Comment (mgorny) [reply setup.py:25]: The code is not used with this  version at all. This change introduces an unnecessary dependency on package that isn't used, and that distributions such as Gentoo don't provide for py3.12 at all (because they're never necessary).

Comment (dee-me-tree-or-love) [reply setup.py:25]: @mgorny has a good catch actually, thanks for spotting. Let's try to make a quick patch to undo this part ➕

Comment (dee-me-tree-or-love) [reply setup.py:25]: resolving this here: https://github.com/pypiserver/pypiserver/pull/627
